src/registrar/models/domain.py

```diff
@@ -267,25 +373,234 @@ def administrative_contact(self) -> PublicContact:
+        # if contacts exist and security contact is in the contact list
+        # return that contact
+        # else call the setter
+        #   send the public default contact
+        try:
+            contacts = self._get_property("contacts")
+            for contact in contacts:
+                ##zander don't do this just to do the bare bones here
+                if "type" in contact.keys() and contact["type"] == PublicContact.ContactTypeChoices.SECURITY:
+                    tempContact= self.get_default_security_contact()
+                    tempContact.email=contact["email"]
+                    return tempContact
+
+        except Exception as err:  # use better error handling
+            logger.info("Couldn't get contact %s" % err)
+
+            # TODO - remove this! ideally it should return None,
+            # but error handling needs to be
+            # added on the security email page so that it can handle it being none
+        return self.get_default_security_contact()
+
+    def _add_registrant_to_existing_domain(self, contact: PublicContact):
+        self._update_epp_contact(contact=contact)
+
+        updateDomain = commands.UpdateDomain(
+            name=self.name, registrant=contact.registry_id
+        )
+        try:
+            registry.send(updateDomain, cleaned=True)
+        except RegistryError as e:
+            logger.error(
+                "Error changing to new registrant error code is %s, error is %s"
+                % (e.code, e)
+            )
+            # TODO-error handling better here?
+
+    def _set_singleton_contact(self, contact: PublicContact, expectedType: str):  # noqa
+        """Sets the contacts by adding them to the registry as new contacts,
+        updates the contact if it is already in epp,
+        deletes any additional contacts of the matching type for this domain
+        does not create the PublicContact object, this should be made beforehand
+        (call save() on a public contact to trigger the contact setters
+        which inturn call this function)
+        Raises ValueError if expected type doesn't match the contact type"""
+        if expectedType != contact.contact_type:
+            raise ValueError(
+                "Cannot set a contact with a different contact type,"
+                " expected type was %s" % expectedType
+            )
+
+        isRegistrant = contact.contact_type == contact.ContactTypeChoices.REGISTRANT
+        isEmptySecurity = (
+            contact.contact_type == contact.ContactTypeChoices.SECURITY
+            and contact.email == ""
+        )
+
+        # get publicContact objects that have the matching
+        # domain and type but a different id
+        # like in highlander we there can only be one
+        hasOtherContact = (
+            PublicContact.objects.exclude(registry_id=contact.registry_id)
+            .filter(domain=self, contact_type=contact.contact_type)
+            .exists()
+        )
```

Comment:
This is being used again in existing contact:
```
            if isEmptySecurity:
                existing_contact = PublicContact.objects.filter(
                    domain=self, contact_type=contact.contact_type
                ).get()
            else:
                existing_contact = (
                    PublicContact.objects.exclude(registry_id=contact.registry_id)
                    .filter(domain=self, contact_type=contact.contact_type)
                    .get()
                )
```
I believe that (and I may be wrong), you can call `.exclude` _after_ `.filter`. If so, then we are doing this filter operation twice unnecessarily. You can create an object which stores the following:
`filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type)` 
This allows you to refactor this check to this (ignore the weird github formatting):
 ```suggestion
        filteredContacts = PublicContact.objects.filter(domain=self, contact_type=contact.contact_type)
        hasOtherContact = (
            filteredContacts.exclude(registry_id=contact.registry_id).exists()
        )
```

More on this below....